File: src/go/src/tao/messagechannel.go
//  File: rpc_channel.go
//  Author: Kevin Walsh <kwalsh@holycross.edu>
//
//  Description: An interface for a Message channel.
//
//  Copyright (c) 2013, Google Inc.  All rights reserved.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

package tao

import (
	"code.google.com/p/goprotobuf/proto"
	"encoding/binary"
	"errors"
	"fmt"
	"io"
	"math"
	"os"
)

const (
	DefaultMaxMessageSize = 20 * 1024 * 1024
)

type MessageChannel struct {
	maxMessageSize uint
	pipe io.ReadWriteCloser
}

// Close closes a channel. It is safe to call this multiple times.
func (mc *MessageChannel) Close() error {
	if mc.pipe != nil {
		mc.pipe.Close()
		mc.pipe = nil
	}
	return nil
}

// IsClosed checks if a channel is closed.
func (mc *MessageChannel) IsClosed() bool {
	return mc.pipe == nil
}

// MaxMessageSize gets the maximum message reception size.
func (mc *MessageChannel) MaxMessageSize() uint {
	return mc.maxMessageSize
}

// SetMaxMessageSize sets the maximum message reception size.
func (mc *MessageChannel) SetMaxMessageSize(size uint) {
	mc.maxMessageSize = size
}

// SendData sends raw data to the channel.
// Failure will close the channel.
func (mc *MessageChannel) SendData(bytes []byte) error {
	_, err := mc.pipe.Write(bytes)
	if err != nil {
		mc.Close()
	}
	return err
}

// ReceiveData receives raw data from the channel.
// No maximum message size applies, the caller is expected to supply a
// reasonable size buffer, which will be filled entirely.
// Failure or eof will close the channel.
func (mc *MessageChannel) ReceiveData(bytes []byte) error {
	_, err := io.ReadFull(mc.pipe, bytes)
	if err != nil {
		mc.Close()
	}
	return err
}

// SendString sends a raw string to the channel.
// Failure will close the channel.
func (mc *MessageChannel) SendString(s string) error {
	n := len(s)
	if n > math.MaxUint32 {
		return errors.New("String too large")
	}
	bytes := make([]byte, 4+n)
	binary.BigEndian.PutUint32(bytes[0:4], uint32(n))
	copy(bytes[4:], s)
	return mc.SendData(bytes)
}

// ReceiveString receives a string over the channel.
// Failure or eof will close the channel.
func (mc *MessageChannel) ReceiveString() (string, error) {
	var sizebytes [4]byte
	_, err := io.ReadFull(mc.pipe, sizebytes[:])
	if err != nil {
		return "", err
	}
	n := binary.BigEndian.Uint32(sizebytes[:])
	strbytes := make([]byte, n)
	_, err = io.ReadFull(mc.pipe, strbytes)
	if err != nil {
		return "", err
	}
	return string(strbytes), nil
}

// SendMessage sends a Message to the channel.
// Failure will close the channel.
func (mc *MessageChannel) SendMessage(m proto.Message) error {
	bytes, err := proto.Marshal(m)
	if err != nil {
		mc.Close()
		return err
	}
	return mc.SendString(string(bytes))
}

// ReceiveMessage receives a Message (of a particular type) over the
// channel. Failure or eof will close the channel.
func (mc *MessageChannel) ReceiveMessage(m proto.Message) error {
	s, err := mc.ReceiveString()
	if err != nil {
		return err
	}
	err = proto.Unmarshal([]byte(s), m)
	if err != nil {
		mc.Close()
		return err
	}
	return nil
}

type readerWriterPair struct {
	io.Reader
	io.Writer
	readCloser, writeCloser io.Closer
}

func (pipe *readerWriterPair) Close() error {
	var err1, err2 error
	if pipe.readCloser != nil {
		err1 = pipe.readCloser.Close()
	}
	if pipe.writeCloser != nil {
		err2 = pipe.writeCloser.Close()
	}
	if err1 != nil {
		return err1
	} else {
		return err2
	}
}

func DeserializeMessageChannel(s string) *MessageChannel {
	var readfd, writefd uintptr
	_, err := fmt.Sscanf(s, "tao::FDMessageChannel(%d, %d)", &readfd, &writefd)
	if err != nil {
		fmt.Printf("bad scanf: %s\n", s)
		return nil
	}
	reader := os.NewFile(readfd, "read pipe")
	writer := os.NewFile(writefd, "write pipe")
	var pipe io.ReadWriteCloser
	if readfd == writefd {
		pipe = &readerWriterPair{reader, writer, reader, nil}
	} else {
		pipe = &readerWriterPair{reader, writer, reader, writer}
	}
	return &MessageChannel{DefaultMaxMessageSize, pipe}
}

File: src/tao/message_channel.cc
//  File: message_channel.cc
//  Author: Tom Roeder <tmroeder@google.com>
//
//  Description: A MessageChannel that communicates over Unix file descriptors.
//
//  Copyright (c) 2013, Google Inc.  All rights reserved.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.
#include "tao/message_channel.h"

#include <arpa/inet.h>
#include <unistd.h>

#include <list>
#include <string>

#include <glog/logging.h>
#include <google/protobuf/message.h>

#include "tao/util.h"

namespace tao {
constexpr size_t MessageChannel::DefaultMaxMessageSize;

bool MessageChannel::SendString(const string &s) {
  uint32_t net_len = htonl(s.size());
  return SendData(&net_len, sizeof(net_len)) && SendData(s.c_str(), s.size());
}

bool MessageChannel::SendMessage(const google::protobuf::Message &m) {
  string serialized;
  if (!m.SerializeToString(&serialized)) {
    LOG(ERROR) << "Could not serialize the Message to a string";
    Close();  // Not really necessary, but simplifies semantics.
    return false;
  }
  return SendString(serialized);
}

bool MessageChannel::ReceiveData(void *buffer, size_t buffer_len, bool *eof) {
  if (IsClosed()) {
    LOG(ERROR) << "Can't receive data, channel is already closed";
    *eof = true;
    return true;
  } else {
    *eof = false;
  }
  size_t filled_len = 0;
  while (filled_len != buffer_len) {
    size_t recv_len;
    if (!ReceivePartialData(
            reinterpret_cast<unsigned char *>(buffer) + filled_len,
            buffer_len - filled_len, &recv_len, eof)) {
      LOG(ERROR) << "Failed to read data";
      return false;
    } else if (*eof && filled_len != 0) {
      LOG(ERROR) << "Failed to read complete data";
      return false;
    } else if (*eof) {
      return true;
    }
    filled_len += recv_len;
  }
  return true;
}

bool MessageChannel::ReceiveString(string *s, bool *eof) {
  uint32_t net_len;
  if (!ReceiveData(&net_len, sizeof(net_len), eof)) {
    LOG(ERROR) << "Could not get the length of the data";
    return false;
  } else if (*eof) {
    return true;
  }
  uint32_t len = ntohl(net_len);
  if (len > MaxMessageSize()) {
    LOG(ERROR) << "Message exceeded maximum allowable size";
    Close();
    return false;
  }
  unique_ptr<char[]> temp_data(new char[len]);
  if (!ReceiveData(temp_data.get(), static_cast<size_t>(len), eof) || *eof) {
    LOG(ERROR) << "Could not get the data";
    return false;
  }
  s->assign(temp_data.get(), len);
  return true;
}

bool MessageChannel::ReceiveMessage(google::protobuf::Message *m, bool *eof) {
  string s;
  if (!ReceiveString(&s, eof)) {
    LOG(ERROR) << "Could not receive message";
    return false;
  } else if (*eof) {
    return true;
  }
  if (!m->ParseFromString(s)) {
    LOG(ERROR) << "Could not parse message";
    Close();
    return false;
  }
  return true;
}

}  // namespace tao
